### PR TITLE
Add documentation of all OpenStack provider opts

### PR DESCRIPTION
**Pull request by xsgordon**
Add documentation of all OpenStack provider options including an
indication as to which items are required versus optional and which
configuration blocks they must appear in.

https://github.com/kubernetes/website/issues/5980

---
This change is [Reviewable](https://reviewable.kubernetes.io/reviews/kubernetes/website/6047)

### CODE REVIEW

**Comment by steveperry-53** [comment]
Does anyone understand why the bot labeled this as "Do not merge"?
@xsgordon Is it your intention that this should not be merged?

**Comment by tengqm** [comment]
@steveperry-53 I believe the "do-not-merge" label was added because the PR has "[WIP]" in its title.

**Comment by xsgordon** [comment]
@steveperry-53 I marked it WIP as there is still a few TODOs marked.

/cc @dims @anguslees @arthur0 @FengyunPan @k8s-sig-openstack-feature-requests

**Comment by xsgordon** [comment]
Still not sure what the valid values for lb-method are.

**Comment by ss22ever** [comment]
cc ss22ever

File: docs/concepts/cluster-administration/cloud-providers.md [hunk 1]
```diff
@@ -58,50 +58,108 @@ The information for the annotations for AWS is taken from the comments on [aws.g
...
+* `lb-version` (Optional): Used to override automatic version detection. Valid
+  values are `v1` or `v2`. Where no value is provided automatic detection will
+  select the highest supported version exposed by the underlying OpenStack
+  cloud.
+* `subnet-id` (Optional): Used to specify the id of the subnet you want to
+  create your loadbalancer on. Can be found at Network > Networks. Click on the
+  respective network to get its subnets.
+* `floating-network-id` (Optional): If specified, will create a floating IP for
+  the load balancer.
+* `lb-method`:
```
Comment (arthur0) [inline docs/concepts/cluster-administration/cloud-providers.md]: lb-method: The algorithm used to distribute load between the members of the pool. The value can be: ROUND_ROBIN (default), LEAST_CONNECTIONS or SOURCE_IP

File: docs/concepts/cluster-administration/cloud-providers.md [hunk 2]
```diff
@@ -58,50 +58,108 @@ The information for the annotations for AWS is taken from the comments on [aws.g
...
-be used when using OpenStack with Kubernetes.
+be used when using OpenStack with Kubernetes. The OpenStack cloud provider
+implementation for Kubernetes supports the use of these OpenStack services from
+the underlying cloud, where available:
+
+Service                  | API Version(s) | Required
+-------------------------+----------------+----------
+Block Storage (Cinder)   | V1†, V2        | No
+Compute (Nova)           | V2             | No
+Identity (Keystone)      | V2, V3         | Yes
```
Comment (arthur0) [inline docs/concepts/cluster-administration/cloud-providers.md]: Keystone V2 will be removed in the next OpenStack release #52563

File: docs/concepts/cluster-administration/cloud-providers.md [hunk 3]
```diff
@@ -58,50 +58,108 @@ The information for the annotations for AWS is taken from the comments on [aws.g
...
+auth-url=https://<keystone_ip>/identity/v3
+tenant-id=c869168a828847f39f7f06edd7305637
 domain-id=2a73b8f597c04551a0fdc8e95544be8a
 
 [LoadBalancer]
-subnet-id=6937f8fa-858d-4bc9-a3a5-18d2c957166a  
+subnet-id=6937f8fa-858d-4bc9-a3a5-18d2c957166a
 ```
 
 #### Global
```
Comment (arthur0) [inline docs/concepts/cluster-administration/cloud-providers.md]: I suppose that's possible to use `username` or `user-id`,  `tenant-name` or `tenant-id`, `domain-name` or `domain-id`. The `trust-id`, `region` and `ca-file` parameters are missing, @dims do you know what these parameters are about?

File: docs/concepts/cluster-administration/cloud-providers.md (hunk 2 above)
Comment (xsgordon) [reply docs/concepts/cluster-administration/cloud-providers.md]: Ack will note this. I imagine like some of the others on our end we will need to keep this support for a while as we don't yet expose any disclaimers about what OpenStack versions a given K8S release will work with. We probably need to start thinking about it though as we move to drop things.

Comment (hogepodge) [reply docs/concepts/cluster-administration/cloud-providers.md]: Marking endpoints as deprecated and scheduled for removal helps send a clear message about maintaining installations and interoperabaility.

File: docs/concepts/cluster-administration/cloud-providers.md [hunk 4]
```diff
@@ -58,50 +58,108 @@ The information for the annotations for AWS is taken from the comments on [aws.g
...
+These configuration options for the OpenStack provider pertain to its global
+configuration and should appear in the `[Global]` section of the `cloud.conf`
+file:
 
-####  Load Balancer
-* `subnet-id`: Used to specify the id of the subnet you want to create your loadbalancer on. Can be found at Network > Networks. Click on the respective network to get its subnets.
+* `username` (Required): Refers to the username of a valid user set in keystone.
+* `password` (Required): Refers to the password of a valid user set in keystone.
+* `auth-url` (Required): The URL of the keystone API used to authenticate. On OpenStack control panels, this can be found at Access and Security > API Access > Credentials.
+* `tenant-id` (Required): Used to specify the id of the project where you want to create your resources.
```
Comment (hogepodge) [inline docs/concepts/cluster-administration/cloud-providers.md]: Should add a comment about how Keystone v3 has changes tenant to project, and in the future this terminology will be changing.

File: docs/concepts/cluster-administration/cloud-providers.md (hunk 2 above)
Comment (xsgordon) [reply docs/concepts/cluster-administration/cloud-providers.md]: Yes but doesn't address the other end of the problem I am referring to. E.g. for a given K8S release (1.8) what versions of OpenStack are supported? Based on the listed API versions at least Liberty but not clear whether or not that is  actually true in practice.

If Keystone will drop V2 in Queens, when will we drop K8S support for OSP releases < Queens? That is the discussion we need to have.

File: docs/concepts/cluster-administration/cloud-providers.md (hunk 4 above)
Comment (xsgordon) [reply docs/concepts/cluster-administration/cloud-providers.md]: Adding a mention, though not clear we got consensus on changing the K8S facing aspect from the discussion in https://github.com/kubernetes/kubernetes/pull/52595 - for now it just maps automatically.

File: docs/concepts/cluster-administration/cloud-providers.md (hunk 3 above)
Comment (arthur0) [reply docs/concepts/cluster-administration/cloud-providers.md]: A **region** is a general division of an OpenStack deployment. 
Although a region does not have a geographical connotation, a deployment can use a geographical name for a region ID, such as us- east. `/v3/regions` (Keystone)
A **trust** represents a user’s (the trustor) authorization to delegate roles to another user (the trustee), and optionally allow the trustee to impersonate the trustor.  `/v3/OS-TRUST/trusts` (Keystone)

Comment (xsgordon) [reply docs/concepts/cluster-administration/cloud-providers.md]: Added `region` and `trust`, `ca-file` still outstanding.

Comment (arthur0) [reply docs/concepts/cluster-administration/cloud-providers.md]: A region can be composed of multiple availability zones and this information is essential for Federation. 
Once the cluster nodes are labeled (automatically) `failure-domain.beta.kubernetes.io/zone=myzone` and it's needed the label `failure-domain.beta.kubernetes.io/region=myregion` too. 
It's possible to have multiple availability zones and only one region per cluster. 

The Cluster API object (federation) uses the information (Region and Zones) in the ClusterStatus.
https://github.com/kubernetes/federation/blob/master/federation/apis/federation/v1beta1/types.go#L86-L92


Suppose you have a federation with this characteristics:

Federation Name: `fed` created in `mydomain.com` 

|   Clusters   |  Cloud Provider  |  Region     | Zone          |
|:------------:|:----------------:|:-----------:|:-------------:|
| fr-cluster   | OpenStack        | eu-west-2   | eu-france     |
| uk-cluster   | AWS              | eu-west-2   | eu-london     |
| usa-cluster  | Google Cloud     | us-central1 | us-central1-a |
| chn-cluster  | Azure            | east-asia   | xangai        |

Creating a federated service `foo` in the namespace `bar`
The DNS entries in de DNS provider of federation has the format:

foo.bar.fed.svc.mydomain.com 
foo.bar.fed.svc.eu-west-2.mydomain.com 
foo.bar.fed.svc.eu-west-2.eu-france.mydomain.com 
foo.bar.fed.svc.eu-west-2.eu-london.mydomain.com 
foo.bar.fed.svc.us-central1.mydomain.com 
foo.bar.fed.svc.us-central1.us-central1-a.mydomain.com 
...

If the `foo` service is unavailable,  a new CNAME entry is created pointing to a next top-level domain, e.g:

* `foo` service is down in the `eu-france` **zone**
`foo.bar.fed.svc.eu-west-2.eu-france.mydomain.com` > `foo.bar.fed.svc.eu-west-2.mydomain.com`

* `foo` service is down in the `eu-west-2` **region** :
`foo.bar.fed.svc.eu-west-2.mydomain.com` > `foo.bar.fed.svc.mydomain.com`

**Pull request merged**